qiskit/primitives/base_estimator.py

```diff
@@ -0,0 +1,223 @@
+
+        Args:
+            circuits: quantum circuits that represents quantum states
+            observables: observables
+            parameters: parameters of quantum circuits.
+                Defaults to `[circ.parameters for circ in circuits]`
+                The indexing is such that `parameters[i, j]` is the j-th formal parameter of
+                `circuits[i]`.
+
+        Raises:
+            QiskitError: for mismatch of circuits and parameters list.
+        """
+        self._circuits = tuple(circuits)
+        self._observables = tuple(observables)
+        if parameters is None:
+            self._parameters = tuple(circ.parameters for circ in self._circuits)
+        else:
+            self._parameters = tuple(ParameterView(par) for par in parameters)
+            if len(self._parameters) != len(self._circuits):
+                raise QiskitError(
+                    f"Different number of parameters ({len(self._parameters)} and "
+                    f"circuits ({len(self._circuits)}"
+                )
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, *exc_info):
+        self.close()
+
+    @abstractmethod
+    def close(self):
+        """Close the session and free resources"""
+        ...
+
+    @property
+    def circuits(self) -> tuple[QuantumCircuit, ...]:
+        """Quantum circuits that represents quantum states.
+
+        Returns:
+            quantum circuits
+        """
+        return self._circuits
+
+    @property
+    def observables(self) -> tuple[SparsePauliOp, ...]:
+        """Observables
+
+        Returns:
+            observables
+        """
+        return self._observables
+
+    @property
+    def parameters(self) -> tuple[ParameterView, ...]:
+        """Parameters of quantum circuits
+
+        Returns:
+            parameters[i, j] is the j-th parameter of the i-th circuit.
+        """
+        return self._parameters
+
+    @abstractmethod
+    def __call__(
+        self,
+        circuits: Sequence[int],
+        observables: Sequence[int],
+        parameters: Sequence[Sequence[float]],
+        **run_options,
+    ) -> EstimatorResult:
+        """Run the estimation of expectation value(s).
+
+        `circuits`, `observables`, and `parameters` should have the same length.
+        The i-th element of the result is the expectation of observable
+
+            obs = self.observables[observables[i]]
+
+        for the state prepared by
+
+            circ = self.circuits[circuits[i]]
+
+        with bound parameters
+
+            values = parameters[i].
+
+        Args:
+            circuits: the list of circuit indices.
+            observables: the list of observable indices.
+            parameters: concrete parameters to be bound.
+            run_options: runtime options used for circuit execution.
+
+        Returns:
+            EstimatorResult: the result of Estimator.
+        """
+        ...
+
+
+EstimatorFactory = Callable[..., BaseEstimator]
```

Comment:
What is this line doing? It just defines a new type but is not used anywhere.

Reply:
We later need factories to inject estimators (or samplers) into algorithms/applications - at least if we want to keep circuits/observables immutable.

However, defining the factory only via the return type doesn't help in this respect. We would need to fix the complete signature of the function that returns an estimator. 

We'll actually need factories that take circuit + observable as well as factories that only take the observable (think e.g. of QAOA where the circuit is constructed from the observable).

I would suggest removing this line for now (same for sampler below) and then add the necessary interfaces later.

Reply:
If we add `EstimatorFactory`, I think we need to add some examples as well.

Reply:
You convinced me, I'll lremove factories